File: benchmarks/json_roundtrip/java/Json_roundtrip.java
public class Json_roundtrip {
    static long parseInt(String s, int[] pos) {
        long val = 0;
        while (pos[0] < s.length() && Character.isDigit(s.charAt(pos[0]))) {
            val = val * 10 + (s.charAt(pos[0]) - '0');
            pos[0]++;
        }
        return val;
    }

    static double parseDecimal1(String s, int[] pos) {
        long whole = parseInt(s, pos);
        pos[0]++; // '.'
        int frac = s.charAt(pos[0]) - '0';
        pos[0]++;
        return whole + frac / 10.0;
    }

    public static void main(String[] args) {
        long n = args.length > 0 ? Long.parseLong(args[0]) : 200_000;

        StringBuilder sb = new StringBuilder((int) (n * 48 + 16));
        sb.append('[');
        for (long i = 0; i < n; i++) {
            if (i > 0) sb.append(',');
            sb.append("{\"id\":").append(i)
              .append(",\"name\":\"item").append(i)
              .append("\",\"value\":").append(String.format("%.1f", i * 0.5))
              .append('}');
        }
        sb.append(']');
        String json = sb.toString();

        int[] pos = {1}; // skip '['
        long idSum = 0;
        long decodedCount = 0;
        while (json.charAt(pos[0]) != ']') {
            pos[0]++; // '{'
            pos[0] += "\"id\":".length();
            long id = parseInt(json, pos);
            pos[0] += ",\"name\":\"item".length();
            parseInt(json, pos); // skip digits in name, not re-checked
            pos[0] += "\",\"value\":".length();
            double value = parseDecimal1(json, pos);
            pos[0]++; // '}'
            if (pos[0] < json.length() && json.charAt(pos[0]) == ',') pos[0]++;

            if (value != id * 0.5) {
                System.err.println("self-check failed: decoded value mismatch for id " + id);
                System.exit(1);
            }
            idSum += id;
            decodedCount++;
        }

        long expectedSum = n * (n - 1) / 2;
        if (idSum != expectedSum || decodedCount != n) {
            System.err.println("self-check failed: id sum or count mismatch");
            System.exit(1);
        }

        System.out.println(idSum);
    }
}

File: benchmarks/json_roundtrip/rust/json_roundtrip.rs
use std::env;
use std::fmt::Write as _;

fn parse_int(bytes: &[u8], pos: &mut usize) -> i64 {
    let mut val: i64 = 0;
    while bytes[*pos].is_ascii_digit() {
        val = val * 10 + (bytes[*pos] - b'0') as i64;
        *pos += 1;
    }
    val
}

fn parse_decimal1(bytes: &[u8], pos: &mut usize) -> f64 {
    let whole = parse_int(bytes, pos);
    *pos += 1; // '.'
    let frac = (bytes[*pos] - b'0') as f64;
    *pos += 1;
    whole as f64 + frac / 10.0
}

fn main() {
    let n: i64 = env::args().nth(1).map(|s| s.parse().unwrap()).unwrap_or(200_000);

    // --- encode ---
    let mut buf = String::with_capacity((n as usize) * 48 + 16);
    buf.push('[');
    for i in 0..n {
        if i > 0 {
            buf.push(',');
        }
        write!(buf, "{{\"id\":{},\"name\":\"item{}\",\"value\":{:.1}}}", i, i, i as f64 * 0.5).unwrap();
    }
    buf.push(']');

    // --- decode ---
    let bytes = buf.as_bytes();
    let mut pos = 1usize; // skip '['
    let mut id_sum: i64 = 0;
    let mut decoded_count: i64 = 0;
    while bytes[pos] != b']' {
        pos += 1; // '{'
        pos += "\"id\":".len();
        let id = parse_int(bytes, &mut pos);
        pos += ",\"name\":\"item".len();
        parse_int(bytes, &mut pos); // skip digits in name, not re-checked
        pos += "\",\"value\":".len();
        let value = parse_decimal1(bytes, &mut pos);
        pos += 1; // '}'
        if bytes[pos] == b',' {
            pos += 1;
        }

        assert_eq!(value, id as f64 * 0.5, "self-check failed: decoded value mismatch for id {}", id);
        id_sum += id;
        decoded_count += 1;
    }

    let expected_sum = n * (n - 1) / 2;
    assert!(id_sum == expected_sum && decoded_count == n, "self-check failed: id sum or count mismatch");

    println!("{}", id_sum);
}
